go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -0,0 +1,1142 @@
+				ra,
+				frrConfig,
+				nodeName,
+				selectedNetworks,
+				matchedNetworks,
+			)
+			if new == nil {
+				// if we got nil, we didn't match any VRF
+				return nil, nil, fmt.Errorf("%w: FRRConfiguration %q selected for node %q has no VRF matching the RouteAdvertisements target VRF or any selected network",
+					errConfig, frrConfig.Name, nodeName)
+			}
+			generated = append(generated, new)
+		}
+		// check that we matched all the selected networks on 'auto'
+		if ra.Spec.TargetVRF == "auto" && !matchedNetworks.HasAll(selectedNetworks.networks...) {
+			return nil, nil, fmt.Errorf("%w: selected FRRConfigurations for node %q don't match all selected networks with target VRF 'auto'", errConfig, nodeName)
+		}
+	}
+
+	return generated, nads, nil
+}
+
+// generateFRRConfiguration generates a FRRConfiguration from a source for a
+// specific node. Also fills matchedNetworks with the networks that have a VRF
+// that matched any router VRF of the FRRConfiguration.
+func (c *Controller) generateFRRConfiguration(
+	ra *ratypes.RouteAdvertisements,
+	source *frrtypes.FRRConfiguration,
+	nodeName string,
+	selectedNetworks *selectedNetworks,
+	matchedNetworks sets.Set[string],
+) *frrtypes.FRRConfiguration {
+	routers := []frrtypes.Router{}
+	advertisements := sets.New(ra.Spec.Advertisements...)
+
+	// go over the source routers
+	for i, router := range source.Spec.BGP.Routers {
+
+		targetVRF := ra.Spec.TargetVRF
+		var matchedVRF, matchedNetwork string
+		var receivePrefixes, advertisePrefixes []string
+
+		// check if this router's VRF matches the target VRF
+		switch {
+		case targetVRF == "auto" && router.VRF == "":
+			// match on default network/VRF, advertise node prefixes and receive
+			// any prefix of default network.
+			matchedVRF = ""
+			matchedNetwork = types.DefaultNetworkName
+			advertisePrefixes = selectedNetworks.hostNetworkSubnets[matchedNetwork]
+			receivePrefixes = selectedNetworks.networkSubnets[matchedNetwork]
+		case targetVRF == "auto":
+			// match router.VRF to network.VRF, advertise node prefixes and
+			// receive any prefix of the matched network
+			matchedVRF = router.VRF
+			matchedNetwork = selectedNetworks.networkVRFs[matchedVRF]
+			advertisePrefixes = selectedNetworks.hostNetworkSubnets[matchedNetwork]
+			receivePrefixes = selectedNetworks.networkSubnets[matchedNetwork]
+		case targetVRF == "":
+			// match on default network/VRF, advertise node prefixes and
+			// receive any prefix of selected networks
+			matchedVRF = ""
+			matchedNetwork = types.DefaultNetworkName
+			advertisePrefixes = selectedNetworks.hostSubnets
+			receivePrefixes = selectedNetworks.subnets
+		default:
+			// match router.VRF to network.VRF, advertise node prefixes and
+			// receive any prefix of selected networks
+			matchedVRF = targetVRF
+			matchedNetwork = selectedNetworks.networkVRFs[matchedVRF]
+			advertisePrefixes = selectedNetworks.hostSubnets
+			receivePrefixes = selectedNetworks.subnets
```

Comment:
so receive prefixes, you only match the internal network subnets? What if the external BGP router is trying to advertise routes to something on the external network for our UDNs, wont that be filtered?

Reply:
So the admin would set its own FRRConfiguration with the needed receive prefixes and then FRR-K8s merges all of it in a single FRR config.